feat: permit running with sops via aqua

NikitaCOEUR opened the pull request:
[aqua ](https://aquaproj.github.io/docs/) is a declarative CLI Version Manager written in Go. You can manage tool versions with YAML.

This tool permit to define different sops version through repositories. (show https://aquaproj.github.io/docs/tutorial/change-version-per-project)

However, to benefit from this tool, it is necessary that the commands executed through Aqua are done from a directory that has (or one of its parents has) an aqua.yaml configuration file with the declaration of these packages. (https://aquaproj.github.io/docs/tutorial/config-path/)

So when i use your extension, without sops installed "globally" (/usr/local/bin for example), i can't decrypt or encrypt file.

This is the first time I'm trying to modify the code of a VSCode extension, and it's really challenging to debug. Additionally, I've never worked with JavaScript before, so there might be things to review. Don't hesitate!

However, I've tried to keep the code to a minimum to enable the functionality of this extension with Sops via Aqua without breaking existing functionality. I hope this works for you!

Thank you.

NikitaCOEUR commented:
Hey @misak113,
Can you take a look at this ?

Thank you.

misak113 commented on src/extension.ts:
```diff
@@ -558,6 +558,15 @@ async function getSopsGeneralOptions() {
+	// Moreover sops via aqua need to be executed from directory inside repository where config file aqua.yaml exists
+	// Ensuring that sops is executed from directory where .sops.yaml is found permit to use sops via aqua in case of a repository
+	// If .sops.yaml is not found, sops is executed from user home directory like before
+	let sopsConfigUri = await findSopsConfigRecursive(uri.with({ path: path.dirname(uri.path) }));
```
I think that what you try to achieve is to run and find the config file .sops.yml in the context of a decrypted file and change CWD to the folder.
The `agua` comments are useful for a description of your intention but in general, it's not related to agua at all.
So please just put the agua-related comments only to the PR description, Maybe to the commit message as an example of usage. But not in js comments. It can be misleading in future.

NikitaCOEUR replied:
Force push commit [1c64602](https://github.com/signageos/vscode-sops/pull/73/commits/1c64602fc4a21a9c9863a03eb10500d6e2f2d294) less comments related to aqua.

misak113 commented on src/extension.ts:
```diff
@@ -558,6 +558,15 @@ async function getSopsGeneralOptions() {
+	// Ensuring that sops is executed from directory where .sops.yaml is found permit to use sops via aqua in case of a repository
+	// If .sops.yaml is not found, sops is executed from user home directory like before
+	let sopsConfigUri = await findSopsConfigRecursive(uri.with({ path: path.dirname(uri.path) }));
+	spawnOptions.cwd = sopsConfigUri ? path.dirname(sopsConfigUri.path) : spawnOptions.cwd;
```
You can use the if statement here to assign a new value instead of ternary. It's a little bit more readable, and you reduce the operation of reassigning the same value here.

NikitaCOEUR replied:
Using if statement instead of ternary in this commit [1c64602](https://github.com/signageos/vscode-sops/pull/73/commits/1c64602fc4a21a9c9863a03eb10500d6e2f2d294)

misak113 commented on src/extension.ts:
```diff
@@ -512,7 +512,7 @@ async function fileExists(uri: vscode.Uri) {
 }
 
-async function getSopsGeneralOptions() {
+async function getSopsGeneralOptions(uri: vscode.Uri) {
```
I'd give it a little more descriptive name for the variable. Like decryptedFileUri or so. uri is just too general and user of the function can be confused. Or put a JSDoc here.

NikitaCOEUR replied:
Rename input variable uri => fileUriToEncryptOrDecrypt in commit [1c64602](https://github.com/signageos/vscode-sops/pull/73/commits/1c64602fc4a21a9c9863a03eb10500d6e2f2d294)

misak113 reviewed (changes requested):
It's missing changelog notes (follow guidelines in CHANGELOG.md) and updated README.md docs.

NikitaCOEUR commented:
> It's missing changelog notes (follow guidelines in CHANGELOG.md) and updated README.md docs.

I have just addressed your reviews and updated the Readme and Changelog!
I hope this suits your expectations.
Thanks again for your feedback

say25 commented:
Ever since the release of 0.9.0 this extension has stopped working for us.

```bash
Could not decrypt SOPS file `[PATH-TO-FILE]`: Could not decrypt file: `[PATH-TO-FILE]`, could not read config file: open /dev/null: The system cannot find the path specified.
```

```json
{
	"Secret": "ENC[AES256_GCM,data:jWiLBwbliq7sMLhzq3CbNBwJl9P1yyposJg1rylLV5LIv4cYl4rwSJD44h/x49iRhQLIqMuCi/pb,iv:DkvJlpu0GXVoLdFgQbd22n8yNTNGqV53jA1thAXlBcE=,tag:7JK0uYpAl4HQzyUMyBYljQ==,type:str]",
	"sops": {
		"kms": [
			{
				"arn": "arn:aws:kms:us-east-1:[ACCOUNT_ID]:key/[GUID-TO-KEY]",
				"created_at": "2023-09-08T15:02:14Z",
				"enc": "AQICAHiwozDaVzIWHQxhGzHP7SgOqkpq6HwEeWNMXhE0BIMN7AFo2wUnjVlFSA4B8JNk8JhbAAAAfjB8BgkqhkiG9w0BBwagbzBtAgEAMGgGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMYZj8t/sZqo90vsxvAgEQgDvjIqTPCn4E8xnrPKDhzudMOcdXWUPKah2SnH/agZeXv4YplhrktH3rid2qQT5D2Ban1ohoJvLPmR3IyQ==",
				"aws_profile": "[AWS-PROFILE-NAME]"
			}
		],
		"gcp_kms": null,
		"azure_kv": null,
		"hc_vault": null,
		"age": null,
		"lastmodified": "2023-10-03T00:02:22Z",
		"mac": "ENC[AES256_GCM,data:4a8LLjq0kScDxePOCsMJIJWhV4uM5lA5nVsj9fYSC6hIUwHi9N5Md/FUolgYpP8/G97qy37f77PRoelRd5HRgh0tpz2SyuIOnEfGmgQEDTh1+YHBTVBIFQg9SQqxT1qyHXcTv0x0FTNgK3n4g7NbOaXzHKr/v5cyzQDGa4CrjdQ=,iv:7d/sy5xbfwJa3shdefJW8xmDfkg67viNOVtL+m7tCdo=,tag:aWvOu30APvT2am+6Exs0Xg==,type:str]",
		"pgp": null,
		"unencrypted_suffix": "_unencrypted",
		"version": "3.7.3"
	}
}
```

NikitaCOEUR commented:
Hi,

Ok, it seems that this part cause your problem :

        // Error related to this issue https://github.com/getsops/sops/issues/884 is thrown otherwise. 
         // This is a workaround until the issue is fixed. 
         sopsGeneralArgs.push('--config', '/dev/null');

What is your operating system ?

say25 commented:
So I have 2 machines, on my Mac the extension works fine on 0.9.0. On Windows it works only on 0.8.0 and fails to work on 0.9.0

NikitaCOEUR commented:
> So I have 2 machines, on my Mac the extension works fine on 0.9.0. On Windows it works only on 0.8.0 and fails to work on 0.9.0

Ok, I try to setup windows environment for testing and find a fix tomorrow. Thank you for your feedback about this problem.

misak113 commented:
Hi, thank you for the feedback. I'll look into it later today once I get to my dusty testing Windows machine.

say25 commented:
Thanks, I'm happy to help where I can. 

Some other minor feedback:
1. I would have opened a GitHub issue for this but unfortunately issues are disabled on this project.
2. It seems like there wasn't a [GitHub release for 0.9.0](https://github.com/signageos/vscode-sops/releases) which made it a little harder to find and compare the changes in this release

vchirikov commented:
Still doesn't work (after #74 too). Workaround - create fork prior this PR and create custom version for yourself. (and install via vsix)

NikitaCOEUR commented:
What is your operating system? I tested on Windows 10 and Ubuntu 22 + Ubuntu 20 (with the binaries sops and age installed via the aqua utility or directly). Can you tell us more about the errors encountered?

say25 commented:
The solution worked for me

vchirikov commented:
windows11

![image](https://github.com/signageos/vscode-sops/assets/40987940/71c2ef6a-9460-4050-adad-d87fdb79acf1)

This workspace/repository doesn't have `.sops.yaml` files at all, and the new 0.9.1 throw the error.

Doesn't happen before `aqua` PRs

NikitaCOEUR commented:
Hum, Ok, I'm looking for .sops.yaml, i'll try to fix that.

NikitaCOEUR commented:
I can't seem to replicate your issue on Windows 11. Here's what I did:

- Installed 'sops' via Chocolatey.
- Placed the 'age' and 'age-key' binaries in 'C:\Users<youruser>\AppData\Local\Microsoft\WindowsApps'.
- Created an empty folder 'C:\Users<youruser>\TEST'.
- Created a VSCode workspace in 'C:\Users<youruser>\TEST'.
- Created a file 'test.yaml' with dummy data:
```yaml
value: test
```
- Generated a fresh 'age' key in 'C:\Users<youruser>\TEST\key.txt'.
- Encrypted the YAML file using the command:

```shell
sops --age age1xnjs50dta29l0h7y87k8k7zre3ju4ut0zv40tvadh807hukzk5nsg8852x -i -e C:\Users\<youruser>\TEST\test.yaml
```

- Downloaded the 'vscode-sops' extension version 0.9.1.
- Configured the 'Sops > Defaults: Age Key File' entry in the settings of @ext:signageos.signageos-vscode-sops.
- Clicked on the file 'test.yaml' at the root of my Workspace.

This decrypted it into a temporary file: '.decrypted~test.yaml' which I was able to modify and save without any issues.
